File: xpenbox-back/src/main/java/org/xpenbox/common/dto/APIResponseDTO.java
package org.xpenbox.common.dto;

import io.quarkus.runtime.annotations.RegisterForReflection;

/**
 * APIResponseDTO is a generic data transfer object for API responses.
 * @param success Indicates if the API call was successful
 * @param message Message providing additional information about the response
 * @param data The actual response data
 * @param statusCode HTTP status code of the response
 * @param timestamp Timestamp indicating when the response was created
 */
@RegisterForReflection
public record APIResponseDTO<T>(
    boolean success,
	String message,
	T data,
	int statusCode,
	Long timestamp
) {
    /**
     * Create a successful API response
     * @param <T> Type of the data
     * @param message Success message
     * @param data Response data
     * @param statusCode HTTP status code
     * @return APIResponseDTO representing a successful response
     */
    public static <T> APIResponseDTO<T> success(String message, T data, int statusCode) {
        return new APIResponseDTO<>(true, message, data, statusCode, System.currentTimeMillis());
    }

    /**
     * Create an error API response
     * @param <T> Type of the data
     * @param message Error message
     * @param statusCode HTTP status code
     * @return APIResponseDTO representing an error response
     */
    public static <T> APIResponseDTO<T> error(String message, int statusCode) {
        return new APIResponseDTO<>(false, message, null, statusCode, System.currentTimeMillis());
    }
 }

File: xpenbox-front/src/app/feature/common/model/api.response.dto.ts
/**
 * Generic API response DTO interface.
 * T represents the type of the data payload.
 * @template T - The type of the data payload.
 * @property {boolean} success - Indicates if the API call was successful.
 * @property {string} message - A message accompanying the response.
 * @property {T} data - The data payload of the response.
 * @property {number} statusCode - The HTTP status code of the response.
 * @property {number} timestamp - The timestamp when the response was generated.
 */
export interface ApiResponseDTO<T> {
    success: boolean;
    message: string;
    data: T;
    statusCode: number;
    timestamp: number;
}